File: backend/src/main/java/com/beeacademy/backend/service/TeacherCourseService.java
package com.beeacademy.backend.service;

import com.beeacademy.backend.dto.request.CreateChapterRequest;
import com.beeacademy.backend.dto.request.CreateCourseRequest;
import com.beeacademy.backend.dto.request.CreateLessonRequest;
import com.beeacademy.backend.dto.request.UpdateChapterRequest;
import com.beeacademy.backend.dto.request.UpdateCourseRequest;
import com.beeacademy.backend.dto.request.UpdateLessonRequest;
import com.beeacademy.backend.dto.response.PageResponse;
import com.beeacademy.backend.dto.response.TeacherChapterResponse;
import com.beeacademy.backend.dto.response.TeacherCourseDetailResponse;
import com.beeacademy.backend.dto.response.TeacherCourseResponse;
import com.beeacademy.backend.dto.response.TeacherLessonResponse;
import com.beeacademy.backend.exception.BusinessException;
import com.beeacademy.backend.exception.ResourceNotFoundException;
import com.beeacademy.backend.model.ApprovalHistory;
import com.beeacademy.backend.model.Category;
import com.beeacademy.backend.model.Chapter;
import com.beeacademy.backend.model.Course;
import com.beeacademy.backend.model.CourseStatus;
import com.beeacademy.backend.model.Lesson;
import com.beeacademy.backend.model.Profile;
import com.beeacademy.backend.repository.ApprovalHistoryRepository;
import com.beeacademy.backend.repository.CategoryRepository;
import com.beeacademy.backend.repository.ChapterRepository;
import com.beeacademy.backend.repository.CourseRepository;
import com.beeacademy.backend.repository.LessonRepository;
import com.beeacademy.backend.repository.ProfileRepository;
import com.beeacademy.backend.security.AuthenticatedUser;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;

/**
 * Nghiệp vụ quản lý khóa học phía Giáo viên (Phase 1 — CRUD, không upload).
 *
 * <p>Thiết kế: mỗi thao tác thêm/sửa/xóa Chapter và Lesson sử dụng
 * {@link ChapterRepository} và {@link LessonRepository} trực tiếp thay vì
 * mutate collection của Course aggregate. Lý do:
 * <ul>
 *   <li>{@code Course.getChapters()} trả unmodifiableList → không thể add/remove.</li>
 *   <li>Sử dụng repository trực tiếp rõ ràng hơn và tránh N+1 khi không cần
 *       load toàn bộ chapters.</li>
 * </ul>
 *
 * <p>Nguyên tắc phân quyền:
 * <ul>
 *   <li>GV chỉ thao tác được khóa học của chính mình (verify teacherId).</li>
 *   <li>Chỉ sửa nội dung khi status ∈ {DRAFT, NEEDS_REVISION}.</li>
 *   <li>Submit: DRAFT/NEEDS_REVISION → PENDING_REVIEW.</li>
 * </ul>
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class TeacherCourseService {

    private final CourseRepository          courseRepository;
    private final CategoryRepository        categoryRepository;
    private final ProfileRepository         profileRepository;
    private final ChapterRepository         chapterRepository;
    private final LessonRepository          lessonRepository;
    private final ApprovalHistoryRepository approvalHistoryRepository;

    // ========================================================================
    // Course CRUD
    // ========================================================================

    /**
     * Tạo khóa học mới ở trạng thái DRAFT.
     * Slug tự động sinh từ title, đảm bảo unique bằng suffix số.
     */
    @Transactional
    public TeacherCourseResponse createCourse(AuthenticatedUser me,
                                               CreateCourseRequest req) {
        Profile  teacher  = loadProfile(me.userId());
        Category category = loadCategory(req.categoryId());

        // Validate: giá khuyến mãi phải nhỏ hơn giá gốc (cross-field validation)
        validateSalePrice(req.salePriceVnd(), req.priceVnd());

        int[] grades = req.grades().stream().mapToInt(Integer::intValue).toArray();
        Course course = Course.createByTeacher(teacher, req.title(), req.description(),
                                               category, grades, req.priceVnd());

        // Đảm bảo slug unique: thêm suffix -2, -3... nếu đã tồn tại
        String baseSlug = course.getSlug();
        String slug = baseSlug;
        int suffix = 2;
        while (courseRepository.findBySlug(slug).isPresent()) {
            slug = baseSlug + "-" + suffix++;
        }

        // Gán salePriceVnd nếu có (factory không nhận field này).
        // Truyền priceVnd=0 vì Course.update() có guard "if (priceVnd > 0)" —
        // 0 sẽ không ghi đè priceVnd đã được set bởi factory ở trên.
        // KHÔNG truyền null vì priceVnd là primitive int — sẽ gây NullPointerException khi unbox.
        if (req.salePriceVnd() != null) {
            course.update(null, null, null, null, 0, req.salePriceVnd(), null);
        }

        Course saved = courseRepository.save(course);
        log.info("GV {} tạo khóa học '{}' ({})", me.userId(), saved.getTitle(), saved.getId());
        return TeacherCourseResponse.fromEntity(saved);
    }

    /** Danh sách khóa học của GV, sắp xếp theo updatedAt DESC. */
    @Transactional(readOnly = true)
    public PageResponse<TeacherCourseResponse> listMyCourses(AuthenticatedUser me,
                                                               Pageable pageable) {
        Specification<Course> spec = (root, q, cb) ->
                cb.equal(root.get("teacher").get("id"), me.userId());
        Page<Course> page = courseRepository.findAll(spec, pageable);
        return PageResponse.of(page, TeacherCourseResponse::fromEntity);
    }

    /** Chi tiết khóa học + chapters + lessons + lịch sử duyệt. */
    @Transactional(readOnly = true)
    public TeacherCourseDetailResponse getCourseDetail(UUID courseId, AuthenticatedUser me) {
        Course course = loadAndVerifyOwner(courseId, me.userId());
        List<ApprovalHistory> history =
                approvalHistoryRepository.findByCourseIdOrderByCreatedAtAsc(courseId);
        return TeacherCourseDetailResponse.fromEntity(course, history);
    }

    /** Cập nhật thông tin cơ bản (chỉ khi DRAFT/NEEDS_REVISION). */
    @Transactional
    public TeacherCourseResponse updateCourse(UUID courseId, AuthenticatedUser me,
                                               UpdateCourseRequest req) {
        Course   course   = loadAndVerifyOwner(courseId, me.userId());
        assertEditable(course);

        // Tính giá hiệu dụng sau khi update để validate cross-field
        Integer effectivePrice     = req.priceVnd()     != null ? req.priceVnd()     : course.getPriceVnd();
        // BUG FIX: null trong request = "giữ nguyên giá KM cũ", KHÔNG phải "xóa giá KM".
        // Để xóa giá KM, frontend gửi clearSalePrice=true hoặc salePriceVnd=0.
        Integer effectiveSalePrice = req.salePriceVnd() != null ? req.salePriceVnd() : course.getSalePriceVnd();

        // Validate: giá khuyến mãi phải nhỏ hơn giá gốc (sau khi tính giá hiệu dụng)
        validateSalePrice(effectiveSalePrice, effectivePrice);

        Category category = req.categoryId() != null ? loadCategory(req.categoryId()) : null;
        int[] grades = req.grades() != null
                ? req.grades().stream().mapToInt(Integer::intValue).toArray() : null;

        course.update(req.title(), req.description(), category,
                      grades,
                      effectivePrice,
                      effectiveSalePrice,
                      null);
        return TeacherCourseResponse.fromEntity(courseRepository.save(course));
    }

    /** Xóa khóa học — chỉ khi DRAFT. */
    @Transactional
    public void deleteCourse(UUID courseId, AuthenticatedUser me) {
        Course course = loadAndVerifyOwner(courseId, me.userId());
        if (course.getStatus() != CourseStatus.DRAFT) {
            throw new BusinessException("CANNOT_DELETE",
                    "Chỉ có thể xóa khóa học ở trạng thái Bản nháp.");
        }
        courseRepository.delete(course);
        log.info("GV {} xóa khóa học {}", me.userId(), courseId);
    }

    /** Nộp khóa học để Admin duyệt. */
    @Transactional
    public TeacherCourseResponse submitForReview(UUID courseId, AuthenticatedUser me) {
        Course course = loadAndVerifyOwner(courseId, me.userId());

        // BUG FIX: load chapters một lần duy nhất — trước đây query 2 lần cùng một kết quả
        List<Chapter> chapters = chapterRepository.findByCourseIdOrderByPositionAsc(courseId);

        // Validate: khóa học phải có ít nhất 1 chương
        if (chapters.isEmpty()) {
            throw new BusinessException("EMPTY_COURSE",
                    "Khóa học phải có ít nhất 1 chương trước khi nộp duyệt.");
        }

        // Validate: mỗi chương phải có ít nhất 1 bài giảng
        boolean anyLessonless = chapters.stream()
                .anyMatch(ch -> lessonRepository.countByChapterId(ch.getId()) == 0);
        if (anyLessonless) {
            throw new BusinessException("EMPTY_CHAPTER",
                    "Mỗi chương phải có ít nhất 1 bài giảng.");
        }

        course.submitForReview();
        Course saved = courseRepository.save(course);
        log.info("GV {} nộp khóa học {} để duyệt", me.userId(), courseId);
        return TeacherCourseResponse.fromEntity(saved);
    }

    // ========================================================================
    // Chapter CRUD
    // Dùng ChapterRepository trực tiếp — KHÔNG mutate Course.chapters (unmodifiable).
    // ========================================================================

    @Transactional
    public TeacherChapterResponse addChapter(UUID courseId, AuthenticatedUser me,
                                              CreateChapterRequest req) {
        Course course = loadAndVerifyOwner(courseId, me.userId());
        assertEditable(course);

        // Tính position tiếp theo nếu không truyền
        int nextPos = chapterRepository.findByCourseIdOrderByPositionAsc(courseId).size() + 1;
        int position = req.position() != null ? req.position() : nextPos;

        Chapter chapter = Chapter.createNew(course, req.title(), req.description(), position);
        Chapter saved   = chapterRepository.save(chapter);
        log.info("Thêm chương '{}' vào khóa học {}", req.title(), courseId);
        return TeacherChapterResponse.fromEntity(saved);
    }

    @Transactional
    public TeacherChapterResponse updateChapter(UUID courseId, UUID chapterId,
                                                 AuthenticatedUser me,
                                                 UpdateChapterRequest req) {
        loadAndVerifyOwner(courseId, me.userId());

        // Verify chapter thuộc đúng courseId
        Chapter chapter = chapterRepository.findByIdAndCourseId(chapterId, courseId)
                .orElseThrow(() -> new ResourceNotFoundException("Chapter", chapterId));

        // assertEditable dựa trên course status — load lại để lấy status
        Course course = courseRepository.findById(courseId)
                .orElseThrow(() -> new ResourceNotFoundException("Course", courseId));
        assertEditable(course);

        chapter.update(req.title(), req.description(), req.position());
        return TeacherChapterResponse.fromEntity(chapterRepository.save(chapter));
    }

    @Transactional
    public void deleteChapter(UUID courseId, UUID chapterId, AuthenticatedUser me) {
        Course course = loadAndVerifyOwner(courseId, me.userId());
        assertEditable(course);

        // Verify chapter thuộc courseId trước khi xóa (tránh xóa nhầm chapter của người khác)
        Chapter chapter = chapterRepository.findByIdAndCourseId(chapterId, courseId)
                .orElseThrow(() -> new ResourceNotFoundException("Chapter", chapterId));

        // CascadeType.ALL + orphanRemoval trên lessons → xóa lessons theo tự động
        chapterRepository.delete(chapter);
        log.info("Xóa chương {} khỏi khóa học {}", chapterId, courseId);
    }

    // ========================================================================
    // Lesson CRUD
    // Dùng LessonRepository trực tiếp — KHÔNG mutate Chapter.lessons (unmodifiable).
    // ========================================================================

    @Transactional
    public TeacherLessonResponse addLesson(UUID courseId, UUID chapterId,
                                            AuthenticatedUser me,
                                            CreateLessonRequest req) {
        Course course = loadAndVerifyOwner(courseId, me.userId());
        assertEditable(course);

        Chapter chapter = chapterRepository.findByIdAndCourseId(chapterId, courseId)
                .orElseThrow(() -> new ResourceNotFoundException("Chapter", chapterId));

        int nextPos = lessonRepository.countByChapterId(chapterId) + 1;
        int position = req.position() != null ? req.position() : nextPos;

        Lesson lesson = Lesson.createNew(chapter, req.title(), req.description(),
                                         position, req.isFree());
        if (req.videoEmbedUrl() != null && !req.videoEmbedUrl().isBlank()) {
            lesson.update(null, null, null, req.isFree(), req.videoEmbedUrl());
        }

        Lesson saved = lessonRepository.save(lesson);
        log.info("Thêm bài giảng '{}' vào chương {}", req.title(), chapterId);
        return TeacherLessonResponse.fromEntity(saved);
    }

    @Transactional
    public TeacherLessonResponse updateLesson(UUID courseId, UUID chapterId,
                                               UUID lessonId, AuthenticatedUser me,
                                               UpdateLessonRequest req) {
        Course course = loadAndVerifyOwner(courseId, me.userId());
        assertEditable(course);

        // Verify ownership chain: lesson → chapter → course → teacher
        chapterRepository.findByIdAndCourseId(chapterId, courseId)
                .orElseThrow(() -> new ResourceNotFoundException("Chapter", chapterId));

        Lesson lesson = lessonRepository.findByIdAndChapterId(lessonId, chapterId)
                .orElseThrow(() -> new ResourceNotFoundException("Lesson", lessonId));

        lesson.update(
                req.title(),
                req.description(),
                req.position(),
                req.isFree() != null ? req.isFree() : lesson.getIsFree(),
                req.videoEmbedUrl());

        return TeacherLessonResponse.fromEntity(lessonRepository.save(lesson));
    }

    @Transactional
    public void deleteLesson(UUID courseId, UUID chapterId, UUID lessonId,
                              AuthenticatedUser me) {
        Course course = loadAndVerifyOwner(courseId, me.userId());
        assertEditable(course);

        chapterRepository.findByIdAndCourseId(chapterId, courseId)
                .orElseThrow(() -> new ResourceNotFoundException("Chapter", chapterId));

        Lesson lesson = lessonRepository.findByIdAndChapterId(lessonId, chapterId)
                .orElseThrow(() -> new ResourceNotFoundException("Lesson", lessonId));

        lessonRepository.delete(lesson);
        log.info("Xóa bài giảng {} khỏi chương {}", lessonId, chapterId);
    }

    // ========================================================================
    // Private helpers
    // ========================================================================

    /**
     * Load course và verify GV là owner.
     * Ném 404 nếu không tồn tại, 403 nếu không phải owner.
     */
    private Course loadAndVerifyOwner(UUID courseId, UUID teacherId) {
        Course course = courseRepository.findWithCategoryAndTeacherById(courseId)
                .orElseThrow(() -> new ResourceNotFoundException("Course", courseId));
        if (!course.getTeacher().getId().equals(teacherId)) {
            throw new BusinessException("FORBIDDEN",
                    "Bạn không có quyền chỉnh sửa khóa học này.", HttpStatus.FORBIDDEN);
        }
        return course;
    }

    /**
     * Chỉ cho phép edit khi status ∈ {DRAFT, NEEDS_REVISION, REJECTED}.
     *
     * <p>Các trạng thái cho phép sửa:
     * <ul>
     *   <li>DRAFT — khóa học mới tạo, chưa nộp.</li>
     *   <li>NEEDS_REVISION — Admin yêu cầu chỉnh sửa rồi nộp lại.</li>
     *   <li>REJECTED — Admin từ chối; GV cần chỉnh sửa rồi nộp lại qua draft mới.</li>
     * </ul>
     *
     * <p>Các trạng thái BỊ CHẶN:
     * <ul>
     *   <li>PENDING_REVIEW — đang chờ Admin duyệt, không được sửa.</li>
     *   <li>APPROVED — Admin đã duyệt, chờ publish.</li>
     *   <li>PUBLISHED — đang hiển thị cho học sinh, không sửa không qua duyệt.</li>
     *   <li>ARCHIVED — đã lưu trữ.</li>
     * </ul>
     */
    private void assertEditable(Course course) {
        CourseStatus s = course.getStatus();
        // Whitelist approach: chỉ cho phép khi status thuộc danh sách an toàn
        boolean editable = s == CourseStatus.DRAFT
                        || s == CourseStatus.NEEDS_REVISION
                        || s == CourseStatus.REJECTED;  // GV cần sửa sau khi bị từ chối
        if (!editable) {
            String statusLabel = switch (s) {
                case PENDING_REVIEW -> "Đang chờ duyệt";
                case APPROVED       -> "Đã duyệt (chờ publish)";
                case PUBLISHED      -> "Đã phát hành";
                case ARCHIVED       -> "Đã lưu trữ";
                default             -> s.toDbValue();
            };
            throw new BusinessException("NOT_EDITABLE",
                    "Không thể chỉnh sửa khi khóa học đang ở trạng thái '"
                    + statusLabel + "'.");
        }
    }

    /**
     * Validate giá khuyến mãi phải nhỏ hơn giá gốc.
     * Bỏ qua nếu salePriceVnd = null (không áp dụng KM).
     */
    private void validateSalePrice(Integer salePriceVnd, Integer priceVnd) {
        if (salePriceVnd == null || priceVnd == null) return;
        if (salePriceVnd >= priceVnd) {
            throw new BusinessException("INVALID_SALE_PRICE",
                    "Giá khuyến mãi (" + salePriceVnd + " VND) phải nhỏ hơn giá gốc ("
                    + priceVnd + " VND).");
        }
        if (salePriceVnd < 1000) {
            throw new BusinessException("INVALID_SALE_PRICE",
                    "Giá khuyến mãi tối thiểu 1,000 VND.");
        }
    }

    private Category loadCategory(UUID id) {
        return categoryRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Category", id));
    }

    private Profile loadProfile(UUID id) {
        return profileRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Profile", id));
    }
}

File: frontend/src/api/teacherCourseService.ts
/**
 * Teacher Course Portal API service
 * Gọi các endpoint /api/teacher/** của backend Spring Boot.
 * Tất cả request tự gắn Bearer token qua apiClient interceptor.
 */
import { apiClient, unwrap } from './client';
import type { ApiResponse, PageResponse } from '../types/api';

// ─── Types ───────────────────────────────────────────────────────────────────

export interface TeacherCourseResponse {
  id: string;
  slug: string;
  title: string;
  thumbnailUrl: string | null;
  /** UUID danh mục — thêm mới để form edit không cần gọi getCourseDetail() thêm lần nữa. */
  categoryId: string | null;
  categoryName: string | null;
  grades: number[];
  priceVnd: number;
  salePriceVnd: number | null;
  status: 'draft' | 'pending_review' | 'approved' | 'rejected' | 'needs_revision' | 'published';
  totalChapters: number;
  totalLessons: number;
  createdAt: string;
  updatedAt: string;
}

export interface TeacherLessonResponse {
  id: string;
  title: string;
  description: string | null;
  position: number;
  isFree: boolean;
  videoEmbedUrl: string | null;
  videoStoragePath: string | null;
  videoUrl: string | null;
  durationSec: number;
  hasVideo: boolean;
}

export interface TeacherChapterResponse {
  id: string;
  title: string;
  description: string | null;
  position: number;
  lessons: TeacherLessonResponse[];
}

export interface TeacherCourseDetailResponse extends TeacherCourseResponse {
  description: string | null;
  categoryId: string | null;
  chapters: TeacherChapterResponse[];
  approvalHistory: ApprovalHistoryResponse[];
}

export interface ApprovalHistoryResponse {
  id: string;
  action: 'approved' | 'rejected' | 'needs_revision';
  comment: string | null;
  adminName: string;
  createdAt: string;
}

export interface CreateCourseRequest {
  title: string;
  description?: string;
  categoryId: string;
  grades: number[];
  priceVnd: number;
  salePriceVnd?: number;
}

export interface CreateChapterRequest {
  title: string;
  description?: string;
  position?: number;
}

export interface CreateLessonRequest {
  title: string;
  description?: string;
  position?: number;
  isFree: boolean;
  videoEmbedUrl?: string;
}

// ─── Course CRUD ─────────────────────────────────────────────────────────────

export async function createCourse(req: CreateCourseRequest): Promise<TeacherCourseResponse> {
  const res = await apiClient.post<ApiResponse<TeacherCourseResponse>>(
    '/api/teacher/courses', req);
  return unwrap(res.data);
}

export async function listMyCourses(page = 0, size = 10):
    Promise<PageResponse<TeacherCourseResponse>> {
  const res = await apiClient.get<ApiResponse<PageResponse<TeacherCourseResponse>>>(
    '/api/teacher/courses', { params: { page, size, sort: 'updatedAt,desc' } });
  return unwrap(res.data);
}

export async function getCourseDetail(courseId: string):
    Promise<TeacherCourseDetailResponse> {
  const res = await apiClient.get<ApiResponse<TeacherCourseDetailResponse>>(
    `/api/teacher/courses/${courseId}`);
  return unwrap(res.data);
}

export async function updateCourse(courseId: string, req: Partial<CreateCourseRequest>):
    Promise<TeacherCourseResponse> {
  const res = await apiClient.put<ApiResponse<TeacherCourseResponse>>(
    `/api/teacher/courses/${courseId}`, req);
  return unwrap(res.data);
}

export async function deleteCourse(courseId: string): Promise<void> {
  await apiClient.delete(`/api/teacher/courses/${courseId}`);
}

export async function submitForReview(courseId: string): Promise<TeacherCourseResponse> {
  const res = await apiClient.post<ApiResponse<TeacherCourseResponse>>(
    `/api/teacher/courses/${courseId}/submit`);
  return unwrap(res.data);
}

// ─── Chapter CRUD ─────────────────────────────────────────────────────────────

export async function addChapter(courseId: string, req: CreateChapterRequest):
    Promise<TeacherChapterResponse> {
  const res = await apiClient.post<ApiResponse<TeacherChapterResponse>>(
    `/api/teacher/courses/${courseId}/chapters`, req);
  return unwrap(res.data);
}

export async function updateChapter(courseId: string, chapterId: string,
                                     req: Partial<CreateChapterRequest>):
    Promise<TeacherChapterResponse> {
  const res = await apiClient.put<ApiResponse<TeacherChapterResponse>>(
    `/api/teacher/courses/${courseId}/chapters/${chapterId}`, req);
  return unwrap(res.data);
}

export async function deleteChapter(courseId: string, chapterId: string): Promise<void> {
  await apiClient.delete(`/api/teacher/courses/${courseId}/chapters/${chapterId}`);
}

// ─── Lesson CRUD ─────────────────────────────────────────────────────────────

export async function addLesson(courseId: string, chapterId: string,
                                  req: CreateLessonRequest): Promise<TeacherLessonResponse> {
  const res = await apiClient.post<ApiResponse<TeacherLessonResponse>>(
    `/api/teacher/courses/${courseId}/chapters/${chapterId}/lessons`, req);
  return unwrap(res.data);
}

export async function updateLesson(courseId: string, chapterId: string,
                                    lessonId: string,
                                    req: Partial<CreateLessonRequest>):
    Promise<TeacherLessonResponse> {
  const res = await apiClient.put<ApiResponse<TeacherLessonResponse>>(
    `/api/teacher/courses/${courseId}/chapters/${chapterId}/lessons/${lessonId}`, req);
  return unwrap(res.data);
}

export async function deleteLesson(courseId: string, chapterId: string,
                                    lessonId: string): Promise<void> {
  await apiClient.delete(
    `/api/teacher/courses/${courseId}/chapters/${chapterId}/lessons/${lessonId}`);
}

// ─── Upload ───────────────────────────────────────────────────────────────────

export interface UploadResponse {
  storagePath: string;
  publicUrl: string | null;
  fileType: string;
  fileSizeBytes: number;
}

export async function uploadVideo(
    courseId: string, chapterId: string, lessonId: string,
    file: File,
    onProgress?: (pct: number) => void): Promise<UploadResponse> {
  const form = new FormData();
  form.append('file', file);
  const res = await apiClient.post<ApiResponse<UploadResponse>>(
    `/api/upload/video/${courseId}/${chapterId}/${lessonId}`, form, {
      headers: { 'Content-Type': 'multipart/form-data' },
      onUploadProgress: (e) => {
        if (onProgress && e.total) onProgress(Math.round(e.loaded / e.total * 100));
      },
    });
  return unwrap(res.data);
}

export async function uploadDocument(lessonId: string, file: File,
                                      displayName?: string): Promise<UploadResponse> {
  const form = new FormData();
  form.append('file', file);
  if (displayName) form.append('name', displayName);
  const res = await apiClient.post<ApiResponse<UploadResponse>>(
    `/api/upload/document/${lessonId}`, form,
    { headers: { 'Content-Type': 'multipart/form-data' } });
  return unwrap(res.data);
}
